### PR TITLE
Add categories

### PR DESCRIPTION
**Pull request by OlexVoit**

### CODE REVIEW

File: src/main/java/mate/academy/bookshop/dto/user/UserRegistrationRequestDto.java [hunk 1]
```diff
@@ -9,7 +9,7 @@
 @Data
 @FieldMatch
 public class UserRegistrationRequestDto {
-    private static final String NOT_BE_NULL = "Don't be null";
+    private static final String NOT_BE_NULL = "Can't be null";
```
Comment (HovorukhaBohdan) [inline src/main/java/mate/academy/bookshop/dto/user/UserRegistrationRequestDto.java:12]: For the future, it is not recommended to post changes to the PR that do not relate to the task. I suggest making a separate request with such changes.

**Review by HovorukhaBohdan — APPROVED** [review]
GJ! Pay attention to my comment :)

File: src/main/java/mate/academy/bookshop/dto/user/UserRegistrationRequestDto.java (hunk 1 above)
Comment (OlexVoit) [reply src/main/java/mate/academy/bookshop/dto/user/UserRegistrationRequestDto.java:12]: ++

File: src/main/java/mate/academy/bookshop/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    public CategoryResponseDto getCategoryById(@PathVariable Long id) {
+        return categoryService.getCategoryById(id);
+    }
+
+    @PreAuthorize("hasAnyRole('ADMIN')")
+    @PutMapping("/{id}")
+    @Operation(summary = "Update category",
+            description = "Only the administrator can change the category")
+    public CategoryResponseDto updateCategory(@PathVariable Long id,
+                                              @RequestBody CreateCategoryRequestDto requestDto) {
```
Comment (vovasalyha) [inline src/main/java/mate/academy/bookshop/controller/CategoryController.java]:
```suggestion
                                              @RequestBody @Valid CreateCategoryRequestDto requestDto) {
```

File: src/main/java/mate/academy/bookshop/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        return categoryService.updateCategory(id, requestDto);
+    }
+
+    @PreAuthorize("hasAnyRole('ADMIN')")
+    @DeleteMapping("/{id}")
+    @Operation(summary = "Delete a category",
+            description = "Only the administrator can delete a category by ID")
+    public void deleteCategory(@PathVariable Long id) {
```
Comment (vovasalyha) [inline src/main/java/mate/academy/bookshop/controller/CategoryController.java:76]:
```suggestion
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void deleteCategory(@PathVariable Long id) {
```

File: src/main/java/mate/academy/bookshop/model/Book.java [hunk 4]
```diff
@@ -41,4 +50,12 @@ public class Book {
 
     @Column(name = "is_deleted", nullable = false)
     private boolean isDeleted = false;
+
+    @EqualsAndHashCode.Exclude
+    @ToString.Exclude
+    @ManyToMany
+    @JoinTable(name = "books-categories",
```
Comment (vovasalyha) [inline src/main/java/mate/academy/bookshop/model/Book.java]: In SQL `_` is more common word separator.
```suggestion
    @JoinTable(name = "books_categories",

```

File: src/main/java/mate/academy/bookshop/model/User.java [hunk 5]
```diff
@@ -54,7 +57,7 @@ public class User implements UserDetails {
     @JoinTable(name = "users_roles",
             joinColumns = @JoinColumn(name = "user_id"),
             inverseJoinColumns = @JoinColumn(name = "role_id"))
-    private Set<Role> roles;
+    private Set<Role> roles = new HashSet<>();
```
Comment (vovasalyha) [inline src/main/java/mate/academy/bookshop/model/User.java:60]: Please don't include unrelated changes to PR. Better make separate PR or directrly push to main if it's something minor

File: src/main/java/mate/academy/bookshop/service/impl/BookServiceImpl.java [hunk 6]
```diff
@@ -17,11 +19,20 @@
...
     private final BookRepository bookRepository;
     private final BookMapper bookMapper;
+    private final CategoryRepository categoryRepository;
 
     @Override
     public BookDto save(CreateBookRequestDto requestDto) {
+        List<Category> categories = categoryRepository
+                .findAllById(requestDto.getCategories());
+        if (categories.isEmpty()) {
+            throw new EntityNotFoundException("There are no categories with id: "
```
Comment (vovasalyha) [inline src/main/java/mate/academy/bookshop/service/impl/BookServiceImpl.java]:
```suggestion
            throw new EntityNotFoundException("There are no categories with IDs: "

```

File: src/main/resources/db/changelog/changes/08-create-categories-table.yaml [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+                    unique: true
+              - column:
+                  name: description
+                  type: VARCHAR(255)
+              - column:
+                  name: is_deleted
+                  type: BOOLEAN
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Comment (vovasalyha) [inline src/main/resources/db/changelog/changes/08-create-categories-table.yaml]: Don't forget about empty lines

**Review by vovasalyha — CHANGES_REQUESTED** [review]
Looks good, but see a couple of comments.

**Review by vovasalyha — APPROVED** [review]
